app/src/org/commcare/dalvik/application/CommCareApplication.java

```diff
@@ -701,12 +702,36 @@ public SQLiteDatabase getHandle() {
         return currentApp.getStorage(name, c);
     }
 
+    public <T extends Persistable> HybridFileBackedSqlStorage<T> getFileBackedAppStorage(String name, Class<T> c) {
+        return currentApp.getFileBackedStorage(name, c);
+    }
+
     public <T extends Persistable> SqlStorage<T> getUserStorage(Class<T> c) {
         return getUserStorage(c.getAnnotation(Table.class).value(), c);
     }
 
     public <T extends Persistable> SqlStorage<T> getUserStorage(String storage, Class<T> c) {
-        return new SqlStorage<>(storage, c, new AndroidDbHelper(this.getApplicationContext()) {
+        return new SqlStorage<>(storage, c, buildUserDbHandle());
+    }
+
+    public <T extends Persistable> HybridFileBackedSqlStorage<T> getFileBackedUserStorage(String storage, Class<T> c) {
+        return new HybridFileBackedSqlStorage<>(storage, c, buildUserDbHandle(), getUserDbDir());
+    }
+
+    /**
+     * Get root directory for user-level file-backed sql storage layer
+     * (currently used for storing fixtures).
+     */
+    public String getUserDbDir() {
+        try {
+            return currentApp.storageRoot() + currentApp + "/" + getSession().getLoggedInUser().getUniqueId() + "/";
+        } catch (SessionUnavailableException e) {
+            return currentApp.storageRoot() + currentApp + "/";

```

Comment:
It worries me a lot that we get the DB in two different ways here depending on whether the session is available, since it feels like a process could try to get the storage and get a fully _different_ location when someone isn't logged in, then log in and the fixture they were parsing is gone. Seems a bit safer to error out here